RFC: Provider default style overrides

ling1726 opened the pull request:
Addresses #23821

[**PREVIEW**](https://github.com/ling1726/fluentui/blob/rfc/provider-style-overrides/rfcs/react-components/convergence/provider-style-overrides.md)

fabricteam commented:
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/656f002f3062c0a7962fb23eb165dec64ecc1509'>656f002f3062c0a7962fb23eb165dec64ecc1509</a></sub>
<div id="bundleSizeComment9423"/>

ling1726 commented on rfcs/react-components/convergence/provider-style-overrides.md:
```diff
@@ -0,0 +1,132 @@
+
+#### Pros and Cons
+
+👍 No extra code changes necessary
+👍 Users will understand what the problem is
+👍 API Consistency (The same pattern must be adopted if user wants their own renderer with RendererProvider)
+
+👎 Extra effort to override one single component
```
```suggestion
- 👍 No extra code changes necessary
 -👍 Users will understand what the problem is
 -👍 API Consistency (The same pattern must be adopted if user wants their own renderer with RendererProvider)

- 👎 Extra effort to override one single component
```

ling1726 commented on rfcs/react-components/convergence/provider-style-overrides.md:
```diff
@@ -0,0 +1,132 @@
+
+#### Pros and Cons
+
+👍 Add styles the same way as the `FluentProvider` currently does it
+👍 No need to know about provider internals
+👍 One time thing to do for `FluenProvider` only if really necessary
+
+👎 slightly different API to `makeStyles`
+👎 something 'special' to do for one component
```
```suggestion
- 👍 Add styles the same way as the `FluentProvider` currently does it
- 👍 No need to know about provider internals
- 👍 One time thing to do for `FluenProvider` only if really necessary

- 👎 slightly different API to `makeStyles`
- 👎 something 'special' to do for one component
```

ling1726 commented on rfcs/react-components/convergence/provider-style-overrides.md:
```diff
@@ -0,0 +1,132 @@
+
+#### Pros and Cons
+
+- 👍 No extra code changes necessary
+ -👍 Users will understand what the problem is
+ -👍 API Consistency (The same pattern must be adopted if user wants their own renderer with RendererProvider)
+
+- 👎 Extra effort to override one single component
```
```suggestion
- 👍 No extra code changes necessary
- 👍 Users will understand what the problem is
- 👍 API Consistency (The same pattern must be adopted if user wants their own renderer with RendererProvider)

- 👎 Extra effort to override one single component
```

andrefcdias reviewed (approved):
I'm leaning towards the `Specific API for overriding provider styles` option as that brings the users an easy way to do what they need to do without having to learn about our internals or the why.
It's only fair that we provide a way to override stuff easily and avoid this problem for people who don't want to follow Fluent strictly, and the less the user needs to read the better.

ling1726 commented on rfcs/react-components/convergence/provider-style-overrides.md:
````diff
@@ -0,0 +1,132 @@
+};
+```
+
+The above example is provlematic because the `TextDirectionProvider` is rendered
````
we would still need `TextDirectionProvider` because there's no guarantee that folks would use logical properties. If we did decide to enforce this then.... I guess we're going full on v10 :)

khmakoto reviewed (commented):
While I think this is something that is needed, I don't think it fully addresses the issues described in #23821.

The first issue talks about if we should have the `lineHeight` style by default or not.

The second issue is actually more broad and refers to where we apply the background of the provider (should it be the element? should it be body? should we not apply it?). Maybe something similar to the `applyTo` prop in v8's `ThemeProvider` is better suited for this.

ling1726 commented:
> While I think this is something that is needed, I don't think it fully addresses the issues described in #23821.
> 
> The first issue talks about if we should have the `lineHeight` style by default or not.
> 
> The second issue is actually broader and refers to where we apply the background of the provider (should it be the element? should it be body? should we not apply it?). Maybe something similar to the `applyTo` prop in v8's `ThemeProvider` is better suited for this.

As @layershifter  mentioned in the original issue, those defaults were agreed on by the design team. In this case why would style overrides not resolve this issue ? It seems to me that whatever defaults we provide we will never satisfy every app. Therefore I don't think the correct focus is on modifying our defaults but letting customers override them

layershifter commented on rfcs/react-components/convergence/provider-style-overrides.md:
```diff
@@ -0,0 +1,132 @@
+
+## Summary
+
+Proposes two methods of overriding the default styles on the `FluentProvider`
```
```suggestion
Proposes two methods of overriding the default styles on the `FluentProvider`.
```

layershifter commented on rfcs/react-components/convergence/provider-style-overrides.md:
```diff
@@ -0,0 +1,132 @@
+The `FluentProvider` ships with [default styles](https://github.com/microsoft/fluentui/blob/656f002f3062c0a7962fb23eb165dec64ecc1509/packages/react-components/react-provider/src/components/FluentProvider/useFluentProviderStyles.ts#L11-L18) which can be too opinionated for some customers, especially those who have
+strict requirements on global styles in their application.
+
+Currently it's not possible to use `makeStyles` to override these default styles. See the example below:
```
It's not entirely true, I suggest to change it:

```suggestion
If `makeStyles` will be used to override these default styles it will result in unexpected issues until additional wrapping will be done. See the example below:
```

layershifter commented on rfcs/react-components/convergence/provider-style-overrides.md:
````diff
@@ -0,0 +1,132 @@
+};
+```
+
+The above example is problematic because the `TextDirectionProvider` is rendered
+internally inside the `FluentProvider`. The provider is required to let `makeStyles` know to resolve and apply RTL
+styles to DOM.
````
```suggestion
`FluentProvider` renders internally `TextDirectionProvider` (from `@griffel/react`). The `TextDirectionProvider` is required to let `makeStyles` know to when to return RTL classes.

In the example above `makeStyles` is called *outside* `FluentProvider` (and `TextDirectionProvider`) thus will have a wrong value for text direction. This problem is not specific to Fluent UI React and can be faced in other React applications.
```

IMO it sounds more accurate.

layershifter commented on rfcs/react-components/convergence/provider-style-overrides.md:
```diff
@@ -0,0 +1,132 @@
+- 👍 Users will understand what the problem is
+- 👍 API Consistency (The same pattern must be adopted if user wants their own renderer with RendererProvider)
+
+- 👎 Extra effort to override one single component
```
Not sure if it's a real problem as it will be done in a limited amount of places

ling1726 replied:
I would say it's still extra effort because it's not obvious and even with documentation it's hard to know to specifically look for it either

layershifter commented on rfcs/react-components/convergence/provider-style-overrides.md:
```diff
@@ -0,0 +1,132 @@
+- 👍 One time thing to do for `FluentProvider` only if really necessary
+
+- 👎 slightly different API to `makeStyles`
+- 👎 something 'special' to do for one component
```
```suggestion
- 👎 Something 'special' to do for one component i.e. one time API
- 👎 Additional handling for build time transforms
```

layershifter commented on rfcs/react-components/convergence/provider-style-overrides.md:
```diff
@@ -0,0 +1,132 @@
+- 👍 No need to know about provider internals
+- 👍 One time thing to do for `FluentProvider` only if really necessary
+
+- 👎 slightly different API to `makeStyles`
```
```suggestion
- 👎 Slightly different API to `makeStyles`
- 👎 `className` prop will be still exposed so users would be still able to pass classes from `makeStyles()` call
```

This is also a bummer as then there is no a single way of doing overrides

layershifter reviewed (commented):
I don't think that new API is a good solution there (I added few more cons to that option).

layershifter commented on rfcs/react-components/convergence/provider-style-overrides.md:
```diff
@@ -0,0 +1,132 @@
+- 👍 API Consistency (The same pattern must be adopted if user wants their own renderer with RendererProvider)
+
+- 👎 Extra effort to override one single component
+
```
```suggestion
- 👍/👎 `TextDirectionProvider` should be re-exported

```

layershifter commented on rfcs/react-components/convergence/provider-style-overrides.md:
```diff
@@ -0,0 +1,135 @@
+The `FluentProvider` ships with [default styles](https://github.com/microsoft/fluentui/blob/656f002f3062c0a7962fb23eb165dec64ecc1509/packages/react-components/react-provider/src/components/FluentProvider/useFluentProviderStyles.ts#L11-L18) which can be too opinionated for some customers, especially those who have
+strict requirements on global styles in their application.
+
+If the Provider is styled using a `makeStyle` call to override these default styles, it can result in unexpected issues since styles will not be RTL safe by default. See the example below:
```
```suggestion
If the FluentProvider is styled using a `makeStyle` call to override these default styles, it can result in unexpected issues since styles will not be RTL safe by default. See the example below:
```

bsunderhus commented:
I'd vote for `Specific API for overriding provider styles`.

I believe that just documenting what will be necessary to override the Provider style will be confusing for some partners, and rather verbose. Not to mention that we're leaking implementation by stating that the Provider basically needs another Provider to properly work in that case.

miroslavstastny commented on rfcs/react-components/convergence/provider-style-overrides.md:
```diff
@@ -0,0 +1,135 @@
+The `FluentProvider` ships with [default styles](https://github.com/microsoft/fluentui/blob/656f002f3062c0a7962fb23eb165dec64ecc1509/packages/react-components/react-provider/src/components/FluentProvider/useFluentProviderStyles.ts#L11-L18) which can be too opinionated for some customers, especially those who have
+strict requirements on global styles in their application.
+
+If the FluentProvider is styled using a `makeStyle` call to override these default styles, it can result in unexpected issues since styles will not be RTL safe by default. See the example below:
```
```suggestion
If the FluentProvider is styled using a `makeStyles` call to override these default styles, it can result in unexpected issues since styles will not be RTL safe by default. See the example below:
```

miroslavstastny reviewed (commented):
I vote for specific API for `FluentProvider` style overrides as it is more understandable.
As `FluentProvider` is a special component (chicken for all the eggs), I think it is OK to have a special API for that.

That being said, I agree with @khmakoto that we still need to think about how to apply the provider styles to a different element than the provider's `div`.

layershifter commented:
@bsunderhus @miroslavstastny just to clarify, the specific API will be a _new_ API, but it will **not** remove/replace the existing one. Are you okay with that?

ling1726 commented:
Updated the RFC to include one single proposal which is what folks seem to be for. Plz re-review

miroslavstastny commented on rfcs/react-components/convergence/provider-style-overrides.md:
````diff
@@ -0,0 +1,138 @@
+};
+```
+
+Additionaly in the `FluentProvider` we can strip and classnames that are known to be created by a standard `makeStyles`
````
```suggestion
Additionaly in the `FluentProvider` we will strip and classnames that are known to be created by a standard `makeStyles`
```
